Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,119 @@
   (with-slots (properties) port
     (setf (getf properties indicator) value)))
 
-(defmethod distribute-event ((port basic-port) event)
-  (dispatch-event (event-sheet event) event))
-
-;;; This method is responsible for synthetising leave/enter events for
-;;; sheets which are not mirrored. Also when the sheet is grabbed we
-;;; redirect events there. This makes "almost compatible" behavior for
-;;; ports which do not implement grabbing *and* works well for sheets
-;;; which are not mirrored. -- jd 2019-08-09
-(defmethod distribute-event :around ((port basic-port) (event pointer-event))
-  (when-let ((sheet (port-grabbed-sheet port)))
-    (unless (typep event '(or pointer-enter-event pointer-exit-event))
-      (if (eq sheet (event-sheet event))
-          (dispatch-event sheet event)
-          (let ((new-event (shallow-copy-object event)))
-            (get-pointer-position (sheet event)
-              (setf (slot-value new-event 'x) x
-                    (slot-value new-event 'y) y
-                    (slot-value new-event 'sheet) sheet))
-            (change-class new-event 'pointer-motion-event)
-            (dispatch-event sheet new-event))))
-    (return-from distribute-event))
+;;; This function determines the sheet to which pointer should be
+;;; delivered. The right thing is not obvious:
+;;;
+;;; - we may assume that event-sheet is set correctly by port
+;;; - we may find the innermost sheet's child and deliver to it
+;;; - we may deliver event to sheet's graft and let it dispatch
+;;;
+;;; Third option would require a default handle-event method to call
+;;; handle-event on its child under the cursor. Such top-down approach
+;;; is appealing if we consider that scrolling pane could filter mouse
+;;; wheel buttons without gross handle-event :around hacks. For now we
+;;; implement the second option with the innermost child. In general
+;;; case we need z-ordering for both strategies. -- jd 2019-08-21
+(defun get-pointer-event-sheet (event &aux (sheet (event-sheet event)))
+  (get-pointer-position (sheet event)
+    (loop
+       for child = (child-containing-position sheet x y)
+       do
+         (when (null child)
+           (return sheet))
+         (multiple-value-setq (x y)
+           (untransform-position (sheet-transformation child) x y))
+         (setf sheet child))))
+
+;;; Function is responsible for making a copy of an immutable event
+;;; and adjusting its coordinates to be in the target-sheet
+;;; coordinates. Optionally it may change event's class.
+(defun dispatch-event-copy (target-sheet event &optional new-class
+                            &aux (sheet (event-sheet event)))
+  (if (eql target-sheet sheet)
+      (dispatch-event sheet event)
+      (let ((new-event (shallow-copy-object event)))
```

Comment:
Can you make `shallow-copy-object` optionally accept a new class and initargs to make this more straightforward?

Reply:
it wouldn't then make a shallow object copy so I don't think it would be a good thing. also I do not understand all implications of passing initargs to allocate-instance so I'd rather keep it as it is.